commit 0c8d84a93e9317793de9ad5167a25aea1f925249 (refs/changes/06/1706/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-03-23T11:10:25-07:00 (2 years, 7 months ago)
    
    MANTA-3198 enable pinger on muskie -> authcache and muskie -> mako connections

diff --git a/main.js b/main.js
index f6933c8..311fb39 100644
--- a/main.js
+++ b/main.js
@@ -217,7 +217,8 @@ function createCueballHttpAgent(cfg) {
         resolvers: sharkCfg.resolvers,
         spares: sharkCfg.spares,
         maximum: sharkCfg.maximum,
-        linger: sharkCfg.maxIdleTime,
+        ping: '/ping',
+        pingInterval: sharkCfg.pingInterval,
         tcpKeepAliveInitialDelay: sharkCfg.maxIdleTime,
         log: sharkCfg.log,
         recovery: {
diff --git a/package.json b/package.json
index bb7cf46..a743a67 100644
--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
         "backoff": "2.3.0",
         "bunyan": "0.22.1",
         "bunyan-syslog": "0.2.2",
-        "cueball": "2.2.0",
+        "cueball": "2.2.3",
         "dashdash": "1.3.2",
         "deep-equal": "0.0.0",
         "dtrace-provider": "0.2.8",
diff --git a/sapi_manifests/muskie/template b/sapi_manifests/muskie/template
index 61d6f7e..bab98dd 100644
--- a/sapi_manifests/muskie/template
+++ b/sapi_manifests/muskie/template
@@ -53,6 +53,25 @@
       "{{AUTH_SERVICE}}"
     ],
 
+    {{!
+      Make sure that if a socket is left idle for >60sec, a dummy "ping"
+      request is made on it. This stops the node.js 2-minute server socket
+      timeout on the authcache side from closing our sockets (and potentially
+      racing with our use of that socket).
+
+      Note that this path doesn't actually have to be handled by the authcache
+      (any non-5xx response code is accepted, e.g. 404 is fine).
+    }}
+    "ping": "/ping",
+    "pingInterval": 60000,
+
+    {{!
+      Separately, we want to enable TCP-level keep-alives on the sockets, to
+      ensure that we notice quickly if an entire CN running an authcache panics
+      or is netsplit from us, to avoid handing it requests that will fail.
+    }}
+    "tcpKeepAliveInitialDelay": 10000,
+
     {{!
       The spares value here should be larger than sharkConfig.spares, by
       a factor probably >2, <10. Mahi connections are shared amongst all reqs
@@ -118,6 +137,9 @@
     "maxIdleTime": 10000,
     "maxClients": 50,
 
+    {{! nginx has much longer max idle times than node.js }}
+    "pingInterval": 14400000,
+
     {{!
       We want spares to be small: for the sake of prudence we don't want lots
       of idle sockets sitting around for no reason. We also want it to be large:
